### ISSUE TITLE
Logstash does not stop after SIGTERM (systemctl stop logstash.service) if Elasticsearch is not available (down or red)

### ISSUE THREAD

**Issue by flyingstar16** [elastic/logstash#12277]
**tl;dr**: after `sudo systemctl stop logstash.service`, logstash, despite having received the SIGTERM, does not stop and, because the systemd unit file is configured with `TimeoutStopSec=infinity`, **hangs a server that's rebooting until a hard reset is performed**.

- Version: 
```
~$ apt-cache policy logstash
logstash:
  Installed: 1:7.9.2-1
  Candidate: 1:7.9.2-1
```
- Operating System: Debian GNU/Linux 10 (buster) - 4.19.0-11-amd64 #1 SMP Debian 4.19.146-1 (2020-09-17) x86_64 GNU/Linux

- Config File (if you have sensitive info, please remove it):
```
$ cat /etc/logstash/logstash.yml | grep -vE '^#'
node.name: <REDACTED>
path.data: /var/lib/logstash
path.logs: /var/log/logstash
```
```
# cat /etc/logstash/startup.options | grep -vE '^#'
LS_HOME=/usr/share/logstash
LS_SETTINGS_DIR=/etc/logstash
LS_OPTS="--path.settings ${LS_SETTINGS_DIR}"
LS_JAVA_OPTS=""
LS_PIDFILE=/var/run/logstash.pid
LS_USER=logstash
LS_GROUP=logstash
LS_GC_LOG_FILE=/var/log/logstash/gc.log
LS_OPEN_FILES=16384
LS_NICE=19
SERVICE_NAME="logstash"
SERVICE_DESCRIPTION="logstash"
```
```
# cat /etc/default/logstash
LS_HOME="/usr/share/logstash"
LS_SETTINGS_DIR="/etc/logstash"
LS_PIDFILE="/var/run/logstash.pid"
LS_USER="logstash"
LS_GROUP="logstash"
LS_GC_LOG_FILE="/var/log/logstash/gc.log"
LS_OPEN_FILES="16384"
LS_NICE="19"
SERVICE_NAME="logstash"
SERVICE_DESCRIPTION="logstash"
```

Please let me know if you need the entirety of my config file(s). It's 479 lines total, and I haven't made any changes to it, nor have I had any problems before

- Sample Data: N/A
- Steps to Reproduce:

Issue `sudo systemctl stop logstash.service`. Logstash receives the SIGTERM but continues to wait for elasticsearch.

This happens when ES is down or still red, and is visible from `systemctl status logstash.service` which states `deactivating (stop-sigterm) since [...]`

![image](https://user-images.githubusercontent.com/582280/94347713-c5b2de80-0036-11eb-8095-54c7522243f8.png)

I know Logstash received the SIGTERM because it's visible in `logstash-plain.log`. I tried an `apt-get --reinstall install logstash` to see if it was an issue on my side, but it keeps happening.

![image](https://user-images.githubusercontent.com/582280/94347650-4de4b400-0036-11eb-99a0-256351eeff46.png)

Please let me know if you need any additional information.

Thanks!

**Comment by ebuildy** [comment]
Maybe there are inflight events in the pipeline, hence Logstash try to output it , but fail, then re-try etc...

Have a look at https://www.elastic.co/guide/en/logstash/current/shutdown.html#shutdown-stall-example

Also, use Logstash HTTP API (port 9600) to see whats happen under the hood:

https://www.elastic.co/guide/en/logstash/current/node-stats-api.html#event-stats

(how many events, hot threads...) this should give you a better idea of the situation!

**Comment by flyingstar16** [comment]
Thanks! That was exactly what was happening; I had a `mutate` filter stuck!

```
[2020-10-10T22:41:52,592][WARN ][org.logstash.execution.ShutdownWatcherExt] {"inflight_count"=>0, "stalling_threads_info"=>{["LogStash::Filters::Mutate", {"add_field"=>{"myfield"=>"generic"}, "add_tag"=>["mytag"], "id"=>"21..."}]=>[{"thread_id"=>39, "name"=>"[main]>worker0", "current_call"=>"[...]/vendo
r/bundle/jruby/2.5.0/gems/stud-0.0.23/lib/stud/interval.rb:89:in `sleep'"}, {"thread_id"=>40, "name"=>"[main]>worker1", "current_call"=>"[...]/vendor/bundle/jruby/2.5.0/gems/stud-0.0.23/lib/stud/interval.rb:89:in `sleep'"}, {"thread_id"=>41, "name"=>"[main]>worker2", "current_call"=>"[...]/vendor/bundle/jruby/2.5.0/gems/stud-0.0.23/lib/stud/interval.rb:89:in `sleep'"}, {"thread
_id"=>42, "name"=>"[main]>worker3", "current_call"=>"[...]/vendor/bundle/jruby/2.5.0/gems/stud-0.0.23/lib/stud/interval.rb:89:in `sleep'"}, {"thread_id"=>43, "name"=>"[main]>worker4", "current_call"=>"[...]/vendor/bundle/jruby/2.5.0/gems/stud-0.0.23/lib/stud/interval.rb:89:in `sleep'"}, {"thread_id"=>44, "name"=>"[main]>worker5", "current_call"=>"[...]/vendor/bundle/jruby/2.5.0
/gems/stud-0.0.23/lib/stud/interval.rb:95:in `sleep'"}, {"thread_id"=>45, "name"=>"[main]>worker6", "current_call"=>"[...]/vendor/bundle/jruby/2.5.0/gems/stud-0.0.23/lib/stud/interval.rb:89:in `sleep'"}, {"thread_id"=>46, "name"=>"[main]>worker7", "current_call"=>"[...]/vendor/bundle/jruby/2.5.0/gems/stud-0.0.23/lib/stud/interval.rb:89:in `sleep'"}]}}
```

For the record (in case someone needs it, although the docs are much better for this), what I did was:
1. turned on [persistent queues](https://www.elastic.co/guide/en/logstash/current/persistent-queues.html) and checked that `/var/lib/logstash/queue` (`path.data/queue`) existed and was doing queueing things
1. stopped Elasticsearch while monitoring Logstash logs
1. when Logstash started complaining about not being able to reach Elasticsearch, I `kill -9`ed it
1. checked `path.data/queue` to see  that the queue data was still there
1. restarted Elasticsearch
1. restarted Logstash while looking at the logs

I was greeted by
```
[2020-10-10T22:43:16,948][WARN ][org.logstash.ackedqueue.Queue] recovered head data page 1 is different than checkpoint, using recovered page information
```

Which told me persistent queues were working as they should have :)

At this point I turned on [unsafe shutdown](https://www.elastic.co/guide/en/logstash/current/shutdown.html#_what_happens_during_a_controlled_shutdown) and repeated the steps above to make sure Logstash "died" during a "normal" systemd stop, which it did!

Thanks again for the suggestion!
